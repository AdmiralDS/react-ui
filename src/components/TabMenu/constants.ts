export type Dimension = 'l' | 'm';

export const TAB_HEIGHT_L = 48;
export const TAB_HEIGHT_M = 40;
export const LINE_HEIGHT = '2px';
export const ICON_SIZE_M = 20;
export const ICON_SIZE_L = 24;
export const TAB_PADDING_M = '0 12px';
export const TAB_PADDING_L = '0 16px';
export const ICON_MARGIN = '8px';
export const BADGE_MARGIN = '8px';
export const OVERFLOW_MENU_CONTAINER_SIZE_M = '40px';
export const OVERFLOW_MENU_CONTAINER_SIZE_L = '48px';
export const OVERFLOW_MENU_MARGIN_M = '10px';
export const OVERFLOW_MENU_MARGIN_L = '12px';
export const FOCUS_BORDER_OFFSET_M = '8px';
export const FOCUS_BORDER_OFFSET_L = '10px';
export const TAB_LEFT_OFFSET_L = -48;
export const TAB_LEFT_OFFSET_M = -40;
