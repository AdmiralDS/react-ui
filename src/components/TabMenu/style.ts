import styled from 'styled-components';

import { typography } from '../Typography';

import type { Dimension } from './constants';
import {
  BADGE_MARGIN,
  ICON_MARGIN,
  ICON_SIZE_L,
  ICON_SIZE_M,
  LINE_HEIGHT,
  TAB_HEIGHT_L,
  TAB_HEIGHT_M,
  TAB_PADDING_L,
  TAB_PADDING_M,
} from './constants';

export const Wrapper = styled.div<{ underline?: boolean; mobile?: boolean }>`
  position: relative;
  display: flex;
  flex: 1 1 auto;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  box-shadow: inset 0 -${LINE_HEIGHT} 0 0 ${({ theme, underline }) => (underline ? theme.color.background.tertiary : 'transparent')};
  overflow-x: ${({ mobile }) => (mobile ? 'scroll' : 'visible')};

  &::-webkit-scrollbar {
    width: 0 !important;
    height: 0 !important;
  }

  overflow: -moz-scrollbars-none;
  -ms-overflow-style: none;
  scrollbar-width: none;
  scrollbar-height: none;
`;

/**
 * Для исправления бага в IE (скачет контент) вместо flex: 1 1 auto используем margin-right: auto
 * Это позволяет избежать скачков контента и при этом ширина TabsWrapper будет равна ровно ширине его контента и не более.
 * Это напрямую влияет на работу observer, повешенного на tabsWrapperRef.current в файле index.ts.
 */
export const TabsWrapper = styled.div`
  display: flex;
  margin-right: auto;
`;

export const Underline = styled.div`
  position: absolute;
  bottom: 0;
  display: flex;
  background-color: ${({ theme }) => theme.color.basic.primary};
  height: ${LINE_HEIGHT};
  transition: all 0.2s ease-out;
`;

export const TabContent = styled.div`
  height: 100%;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
`;

export const TabContentWrapper = styled.span<{ dimension: Dimension }>`
  position: relative;
  display: inline-flex;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: ${({ dimension }) => (dimension === 'l' ? TAB_PADDING_L : TAB_PADDING_M)};
`;

export const Tab = styled.button<{ dimension: Dimension; selected: boolean }>`
  position: relative;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  height: ${({ dimension }) => (dimension === 'm' ? TAB_HEIGHT_M : TAB_HEIGHT_L)}px;
  padding: 0;
  background: transparent;
  appearance: none;
  border: none;
  ${({ dimension }) => (dimension === 'm' ? typography['Body/Body 2 Long'] : typography['Body/Body 1 Long'])}
  ${({ selected, theme }) => !selected && `color: ${theme.color.text.secondary};`}
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;

  &:focus:not(:active) {
    &:before {
      position: absolute;
      content: '';
      border: ${LINE_HEIGHT} solid ${({ theme }) => theme.color.basic.hover};
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
    }
  }
  &:focus,
  & > ${TabContentWrapper} {
    outline: none;
  }

  & svg {
    display: inline-block;
    flex: 1 0 auto;
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.text.secondary};
    }
    width: ${({ dimension }) => (dimension === 'm' ? ICON_SIZE_M : ICON_SIZE_L)}px;
    height: ${({ dimension }) => (dimension === 'm' ? ICON_SIZE_M : ICON_SIZE_L)}px;
    margin-right: ${ICON_MARGIN};
  }
  & [data-badge] {
    margin-left: ${BADGE_MARGIN};
  }

  &:hover:not(:disabled) {
    color: ${({ theme }) => theme.color.basic.hover};
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.basic.hover};
    }
    [data-badge] {
      background: ${({ theme }) => theme.color.basic.hover};
      color: ${({ theme }) => theme.color.text.staticWhite};
    }
  }
  &:active:not(:disabled) {
    color: ${({ theme }) => theme.color.basic.press};
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.basic.press};
    }
    [data-badge] {
      background: ${({ theme }) => theme.color.basic.press};
      color: ${({ theme }) => theme.color.text.staticWhite};
    }
  }
  &:disabled {
    color: ${({ theme }) => theme.color.text.tertiary};
    cursor: default;
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.text.tertiary};
    }
  }
`;
