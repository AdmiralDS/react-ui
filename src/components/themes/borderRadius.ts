import type { Shape } from '#src/components/themes/common';
export const ALL_BORDER_RADIUS_VALUES = [
  'Border radius 0',
  'Border radius 2',
  'Border radius 4',
  'Border radius 6',
  'Border radius 8',
  'Border radius 10',
] as const;

export function smallGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '2px';
    case 'Border radius 4':
      return '4px';
    case 'Border radius 6':
      return '4px';
    case 'Border radius 8':
      return '4px';
    case 'Border radius 10':
      return '4px';
    default:
      return '4px';
  }
}

export function mediumGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '2px';
    case 'Border radius 4':
      return '4px';
    case 'Border radius 6':
      return '6px';
    case 'Border radius 8':
      return '8px';
    case 'Border radius 10':
      return '10px';
    default:
      return '4px';
  }
}

export function largeGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '4px';
    case 'Border radius 4':
      return '8px';
    case 'Border radius 6':
      return '12px';
    case 'Border radius 8':
      return '16px';
    case 'Border radius 10':
      return '20px';
    default:
      return '8px';
  }
}
