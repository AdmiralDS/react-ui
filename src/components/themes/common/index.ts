import { Color } from '#src/components/themes/common/color';
import { ZIndex } from '#src/components/themes/common/zIndex';
import baseStyled, { ThemedStyledInterface } from 'styled-components';

export * from './color';

export interface Shadow {
  ClickableActive: string;
  ClickableDefault: string;
  ClickableHover: string;
  NonClickable: string;
}

export interface Theme {
  color: Color;
  shadow: Shadow;
  zIndex: ZIndex;
  name: 'light' | 'dark';
}

export const styled = baseStyled as ThemedStyledInterface<Theme>;
