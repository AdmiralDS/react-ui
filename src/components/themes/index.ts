export * from './light';
export * from './dark';
export type { Theme } from './common';
