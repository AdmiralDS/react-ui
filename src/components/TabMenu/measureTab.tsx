import ReactDOM from 'react-dom';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import { Badge } from '../Badge';
import { LIGHT_THEME } from '../themes';

import { Tab, TabContent, TabContentWrapper } from './style';
import type { Dimension } from './constants';

const measureElement = ({ icon, badge, content }: any, dimension: Dimension, callback: (width: number) => void) => {
  // Creates the hidden div appended to the document body
  const container = document.createElement('div');
  container.style.position = 'absolute';
  container.style.visibility = 'hidden';
  container.style.left = '-10000';
  container.style.top = '-10000';
  container.id = 'tab-wrapper';
  document.body.appendChild(container);

  const element = (
    <ThemeProvider theme={LIGHT_THEME}>
      <Tab dimension={dimension} selected>
        <TabContentWrapper dimension={dimension} tabIndex={-1}>
          {icon && icon}
          <TabContent>{content}</TabContent>
          {typeof badge !== 'undefined' && (
            <Badge data-badge dimension="s" appearance="lightDisable">
              {badge}
            </Badge>
          )}
        </TabContentWrapper>
      </Tab>
    </ThemeProvider>
  );

  // Renders the React element into the hidden div
  ReactDOM.render(element, container, () => {
    // const width = container.clientWidth;
    const width = container.getBoundingClientRect().width;
    callback(width);
    // remove hidden container from dom
    ReactDOM.unmountComponentAtNode(container);
    container.parentNode?.removeChild(container);
  });
};

export default measureElement;
