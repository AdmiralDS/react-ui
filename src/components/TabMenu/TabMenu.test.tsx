import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { LIGHT_THEME } from '../themes';
import { TabMenu, TabMenuProps } from '../TabMenu';

/**
 * В компоненте TabMenu для вычисления ширины всего компонента используется метод getBoundingClientRect().width,
 * для вычисления размеров отдельных табов - метод node.clientWidth.
 * В случае, если пропс mobile равен true, также используется метод scrollIntoView.
 * Для всех этих методов необходимо писать заглушки (mocks), в противном случае scrollIntoView не сработает,
 * а getBoundingClientRect и clientWidth будут возвращать нулевые значения.
 *
 * Вот ссылки с описанием подобных проблем:
 * https://stackoverflow.com/questions/38656541/change-element-size-using-jest
 * https://github.com/jsdom/jsdom/issues/135#issuecomment-68191941
 * Насколько я поняла Jest "запускает тесты с помощью поддельной реализации DOM" (через jsdom),
 * но jsdom справляется далеко не со всеми методами dom-элементов.
 *
 * Если не настраивать заглушки, то по умолчанию компонент отрендерится с нулевой шириной и все табы будут спрятаны в меню.
 */

// начальные значения свойств
const originalClientWidth: PropertyDescriptor =
  Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'clientWidth') || {};
const originalGetBoundingClientRect =
  Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'getBoundingClientRect') || {};
const originalScrollIntoView =
  Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'scrollIntoView') || (() => undefined);

// устанавливаем заглушки
beforeAll(() => {
  Object.defineProperties(window.HTMLElement.prototype, {
    clientWidth: {
      get: function () {
        // устаналиваю clientWidth в 150 только для табов (tagName и id прописаны в measureTab.ts).
        return this.tagName === 'DIV' && this.id === 'tab-wrapper' ? 150 : 0;
      },
    },
    getBoundingClientRect: {
      // если установить в true, то можно будет перезадавать это свойство в отдельных тестах
      configurable: true,
      get: function () {
        return this.id === 'tabmenu'
          ? () => ({
              // устанавливаю ширину компонента в 1000, чтобы всем табам хватило места
              width: 1000,
              height: 0,
              x: 0,
              y: 0,
              top: 0,
              left: 0,
              right: 0,
              bottom: 0,
              toJSON: () => undefined,
            })
          : () => ({
              width: 0,
              height: 0,
              x: 0,
              y: 0,
              top: 0,
              left: 0,
              right: 0,
              bottom: 0,
              toJSON: () => undefined,
            });
      },
    },
  });

  Object.defineProperty(window.HTMLElement.prototype, 'scrollIntoView', {
    value: () => undefined,
    configurable: true,
    writable: true,
  });
});

// возвращаем начальные значения
afterAll(() => {
  Object.defineProperty(HTMLElement.prototype, 'clientWidth', originalClientWidth);
  Object.defineProperty(HTMLElement.prototype, 'getBoundingClientRect', originalGetBoundingClientRect);
  Object.defineProperty(HTMLElement.prototype, 'scrollIntoView', originalScrollIntoView);
});

describe('TabMenu', () => {
  const tabs = [
    {
      id: '1',
      content: 'Option 1',
      'data-testid': '1',
      className: 'tag',
    },
    {
      id: '2',
      content: 'Option 22',
      className: 'tag',
    },
    {
      id: '3',
      content: 'Option 333',
      className: 'tag',
    },
  ];
  const Component: React.FC<TabMenuProps> = (props) => {
    return (
      <ThemeProvider theme={LIGHT_THEME}>
        <TabMenu {...props} activeTab="1" tabs={tabs} id="tabmenu" />
      </ThemeProvider>
    );
  };

  // test #1
  it('should render default component', () => {
    const onChange = jest.fn();
    const wrapper = render(<Component activeTab="1" onChange={onChange} tabs={tabs} />);
    expect(wrapper).toMatchSnapshot();
  });

  // test #2
  it('should render mobile component', () => {
    /**
     * Все табы будут выведены в одну строку без меню. Если размеры компонента в тесте не важны,
     * то заглушки для  getBoundingClientRect и clientWidth прописывать необязательно, обработчик onChange все равно будет вызван
     */
    const mobileOnChange = jest.fn();
    const wrapper = render(<Component activeTab="1" onChange={mobileOnChange} tabs={tabs} mobile />);
    fireEvent.click(wrapper.getAllByRole('tab')[0]);
    expect(mobileOnChange).toBeCalledTimes(1);
  });

  // test #3
  it('should call onChange when user clicks on first tab', () => {
    const onChange = jest.fn();
    const wrapper = render(<Component activeTab="1" onChange={onChange} tabs={tabs} />);
    fireEvent.click(wrapper.getByTestId('1'));
    expect(onChange).toBeCalledTimes(1);
  });
});
