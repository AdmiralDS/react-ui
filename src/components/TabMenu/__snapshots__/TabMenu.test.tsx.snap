// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabMenu should render default component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  box-shadow: inset 0 -2px 0 0 transparent;
  overflow-x: visible;
  overflow: -moz-scrollbars-none;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -webkit-scrollbar-height: none;
  -moz-scrollbar-height: none;
  -ms-scrollbar-height: none;
  scrollbar-height: none;
}

.c0::-webkit-scrollbar {
  width: 0 !important;
  height: 0 !important;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: auto;
}

.c1 {
  position: absolute;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #0D69F2;
  height: 2px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
}

.c6 {
  height: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 0 16px;
}

.c3 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  padding: 0;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  font-family: 'VTB Group UI',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  font-feature-settings: 'tnum' on,'lnum' on;
  color: #2B313B;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.c3:focus:not(:active):before {
  position: absolute;
  content: '';
  border: 2px solid #0A54C2;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

.c3:focus,
.c3 > .c4 {
  outline: none;
}

.c3 svg {
  display: inline-block;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  width: 24px;
  height: 24px;
  margin-right: 8px;
}

.c3 svg *[fill^='#'] {
  fill: #626F84;
}

.c3 [data-badge] {
  margin-left: 8px;
}

.c3:hover:not(:disabled) {
  color: #0A54C2;
}

.c3:hover:not(:disabled) *[fill^='#'] {
  fill: #0A54C2;
}

.c3:hover:not(:disabled) [data-badge] {
  background: #0A54C2;
  color: #FFFFFF;
}

.c3:active:not(:disabled) {
  color: #0949AA;
}

.c3:active:not(:disabled) *[fill^='#'] {
  fill: #0949AA;
}

.c3:active:not(:disabled) [data-badge] {
  background: #0949AA;
  color: #FFFFFF;
}

.c3:disabled {
  color: #A7B0BE;
  cursor: default;
}

.c3:disabled *[fill^='#'] {
  fill: #A7B0BE;
}

.c7 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  padding: 0;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  font-family: 'VTB Group UI',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  font-feature-settings: 'tnum' on,'lnum' on;
  color: #2B313B;
  color: #626F84;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.c7:focus:not(:active):before {
  position: absolute;
  content: '';
  border: 2px solid #0A54C2;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

.c7:focus,
.c7 > .c4 {
  outline: none;
}

.c7 svg {
  display: inline-block;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  width: 24px;
  height: 24px;
  margin-right: 8px;
}

.c7 svg *[fill^='#'] {
  fill: #626F84;
}

.c7 [data-badge] {
  margin-left: 8px;
}

.c7:hover:not(:disabled) {
  color: #0A54C2;
}

.c7:hover:not(:disabled) *[fill^='#'] {
  fill: #0A54C2;
}

.c7:hover:not(:disabled) [data-badge] {
  background: #0A54C2;
  color: #FFFFFF;
}

.c7:active:not(:disabled) {
  color: #0949AA;
}

.c7:active:not(:disabled) *[fill^='#'] {
  fill: #0949AA;
}

.c7:active:not(:disabled) [data-badge] {
  background: #0949AA;
  color: #FFFFFF;
}

.c7:disabled {
  color: #A7B0BE;
  cursor: default;
}

.c7:disabled *[fill^='#'] {
  fill: #A7B0BE;
}

<body>
    <div>
      <div
        class="c0"
        id="tabmenu"
        role="tablist"
      >
        <div
          aria-hidden="true"
          class="c1"
        />
        <div
          class="c2"
        >
          <button
            aria-selected="true"
            class="c3 tag"
            data-testid="1"
            id="1"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="c4 c5"
              tabindex="-1"
            >
              <div
                class="c6"
              >
                Option 1
              </div>
            </span>
          </button>
          <button
            aria-selected="false"
            class="c7 tag"
            id="2"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="c4 c5"
              tabindex="-1"
            >
              <div
                class="c6"
              >
                Option 22
              </div>
            </span>
          </button>
          <button
            aria-selected="false"
            class="c7 tag"
            id="3"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="c4 c5"
              tabindex="-1"
            >
              <div
                class="c6"
              >
                Option 333
              </div>
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": .c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  box-sizing: border-box;
  box-shadow: inset 0 -2px 0 0 transparent;
  overflow-x: visible;
  overflow: -moz-scrollbars-none;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -webkit-scrollbar-height: none;
  -moz-scrollbar-height: none;
  -ms-scrollbar-height: none;
  scrollbar-height: none;
}

.c0::-webkit-scrollbar {
  width: 0 !important;
  height: 0 !important;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: auto;
}

.c1 {
  position: absolute;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #0D69F2;
  height: 2px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
}

.c6 {
  height: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 0 16px;
}

.c3 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  padding: 0;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  font-family: 'VTB Group UI',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  font-feature-settings: 'tnum' on,'lnum' on;
  color: #2B313B;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.c3:focus:not(:active):before {
  position: absolute;
  content: '';
  border: 2px solid #0A54C2;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

.c3:focus,
.c3 > .c4 {
  outline: none;
}

.c3 svg {
  display: inline-block;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  width: 24px;
  height: 24px;
  margin-right: 8px;
}

.c3 svg *[fill^='#'] {
  fill: #626F84;
}

.c3 [data-badge] {
  margin-left: 8px;
}

.c3:hover:not(:disabled) {
  color: #0A54C2;
}

.c3:hover:not(:disabled) *[fill^='#'] {
  fill: #0A54C2;
}

.c3:hover:not(:disabled) [data-badge] {
  background: #0A54C2;
  color: #FFFFFF;
}

.c3:active:not(:disabled) {
  color: #0949AA;
}

.c3:active:not(:disabled) *[fill^='#'] {
  fill: #0949AA;
}

.c3:active:not(:disabled) [data-badge] {
  background: #0949AA;
  color: #FFFFFF;
}

.c3:disabled {
  color: #A7B0BE;
  cursor: default;
}

.c3:disabled *[fill^='#'] {
  fill: #A7B0BE;
}

.c7 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  padding: 0;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  font-family: 'VTB Group UI',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  font-feature-settings: 'tnum' on,'lnum' on;
  color: #2B313B;
  color: #626F84;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.c7:focus:not(:active):before {
  position: absolute;
  content: '';
  border: 2px solid #0A54C2;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

.c7:focus,
.c7 > .c4 {
  outline: none;
}

.c7 svg {
  display: inline-block;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  width: 24px;
  height: 24px;
  margin-right: 8px;
}

.c7 svg *[fill^='#'] {
  fill: #626F84;
}

.c7 [data-badge] {
  margin-left: 8px;
}

.c7:hover:not(:disabled) {
  color: #0A54C2;
}

.c7:hover:not(:disabled) *[fill^='#'] {
  fill: #0A54C2;
}

.c7:hover:not(:disabled) [data-badge] {
  background: #0A54C2;
  color: #FFFFFF;
}

.c7:active:not(:disabled) {
  color: #0949AA;
}

.c7:active:not(:disabled) *[fill^='#'] {
  fill: #0949AA;
}

.c7:active:not(:disabled) [data-badge] {
  background: #0949AA;
  color: #FFFFFF;
}

.c7:disabled {
  color: #A7B0BE;
  cursor: default;
}

.c7:disabled *[fill^='#'] {
  fill: #A7B0BE;
}

<div>
    <div
      class="c0"
      id="tabmenu"
      role="tablist"
    >
      <div
        aria-hidden="true"
        class="c1"
      />
      <div
        class="c2"
      >
        <button
          aria-selected="true"
          class="c3 tag"
          data-testid="1"
          id="1"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="c4 c5"
            tabindex="-1"
          >
            <div
              class="c6"
            >
              Option 1
            </div>
          </span>
        </button>
        <button
          aria-selected="false"
          class="c7 tag"
          id="2"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="c4 c5"
            tabindex="-1"
          >
            <div
              class="c6"
            >
              Option 22
            </div>
          </span>
        </button>
        <button
          aria-selected="false"
          class="c7 tag"
          id="3"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="c4 c5"
            tabindex="-1"
          >
            <div
              class="c6"
            >
              Option 333
            </div>
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
